File: testdata/fixtures/typescript/CWE-434.ts
// CWE-434: Unrestricted Upload — multer file without allowlist reaches file write/mv
import fs from "fs";
import multer from "multer";
import { Request, Response } from "express";

const upload = multer({ dest: "uploads/" });

export function vuln1(req: Request, res: Response) {
  const f = req.file;
  // ruleid: typescript-434-upload
  fs.rename(f.path, "/uploads/" + f.originalname, () => {});
}

export function vuln2(req: Request, res: Response) {
  const name = req.file.originalname;
  // ruleid: typescript-434-upload
  fs.writeFile("/tmp/" + name, Buffer.from("x"), () => {});
}

export function vuln3(req: Request, res: Response) {
  const file = req.file;
  // ruleid: typescript-434-upload
  file.mv("/uploads/" + file.originalname);
}

export function safeAllowlist(req: Request, res: Response) {
  const f = req.file;
  const ext = f.originalname.split(".").pop();
  if (!["png","jpg","jpeg"].includes(ext)) return res.status(400).send("bad");
  // ok: typescript-434-upload
  fs.rename(f.path, "/uploads/allowed.png", () => {});
}

export function safeFixed(req: Request, res: Response) {
  // ok: typescript-434-upload
  fs.writeFile("/tmp/fixed.png", Buffer.from("fixed"), () => {});
}

File: testdata/fixtures/go/CWE-434.go
// CWE-434: Unrestricted upload — FormFile filename reaches file write.
package main

import (
    "io"
    "io/ioutil"
    "net/http"
    "os"
)

func vuln1(w http.ResponseWriter, r *http.Request) {
    file, header, _ := r.FormFile("file")
    defer file.Close()
    // ruleid: go-unrestricted-upload
    ioutil.WriteFile("/tmp/"+header.Filename, header.Filename, 0644)
    // ruleid: go-unrestricted-upload
    os.Create("/tmp/" + header.Filename)
}

func vuln2(w http.ResponseWriter, r *http.Request) {
    _, hdr, _ := r.FormFile("upload")
    name := hdr.Filename
    // ruleid: go-unrestricted-upload
    os.WriteFile("/uploads/"+name, []byte("x"), 0644)
}

func safeFixed(w http.ResponseWriter, r *http.Request) {
    // ok: go-unrestricted-upload
    os.Create("/tmp/fixed.txt")
    // ok: go-unrestricted-upload
    ioutil.WriteFile("/tmp/fixed.txt", []byte("fixed"), 0644)
}

func safeWhitelist(w http.ResponseWriter, r *http.Request) {
    _, hdr, _ := r.FormFile("file")
    name := hdr.Filename
    if name != "allowed.png" {
        http.Error(w, "bad", 400)
        return
    }
    // ok: go-unrestricted-upload
    os.WriteFile("/tmp/allowed.png", []byte("x"), 0644)
}

// helper to avoid unused import
var _ = io.Copy
